rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+        storage.counter.store(incremented);
+
+        incremented
+    }
+}
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The compiler will handle storage slots assignment for each storage variable. Each storage access expression will simply return a `StorageHandler` containing the key chosen by the compiler.
+
+## Storage Slot Assignment
+
+Each storage variable in a `storage` block is assigned an index corresponding to its location in the list of variables:
+
+```rust
+struct MyInnerStruct {
+    c: u64,
+    d: b256,
+}
+
+struct MyStruct {
+    a: u64,
+    b: MyInnerStruct,
+}
+
+storage {
+    x: u64 = 0, // Assigned index `0`
+    y: b256 = ZERO_B256, // Assigned index `1`
+    str: str[99], // Assigned index '2'
+    s: MyStruct = MyStruct { a: 0, b: MyInnerStruct { c: 0, d: ZERO_B256 } }, // Assigned index `3`
+}
+```
+
+The "index" assigned to each variable is used to generated a key as follows.
+
+### "Small" Types
+
+Small types are types that fit in a single storage slot. We exclude structs with multiple fields from this definition even they do fit in one slot; structs with at least one field are considered "large types" are handled in the [next section](#large-types).
+
+Storage variables that have small types are assigned a single key that is equal to `sha256("storage_<idx>")` where `<idx>` is the index assigned to the storage variable. In the example above, the key chosen for `x` is:
+
+```rust
+sha256("storage_0") = 0xf383b0ce51358be57daa3b725fe44acdb2d880604e367199080b4379c41bb6ed
+```
+
+and the key chosen for `y` is:
+
+```rust
+sha256("storage_1") = 0xde9090cb50e71c2588c773487d1da7066d0c719849a7e58dc8b6397a25c567c0
+```
+
+### "Large" Types
+
+Large types span multiple storage slots and are laid out sequentially in storage starting at key `sha256("storage_<idx>")` where `<idx>` is the index assigned to each storage variable. The number of storage slots required can be computed as `(__size_of::<T>() + 31) >> 5` (ceiling of the size of the data type divided by `32`). Structs are the only exception to this rule because the fields of a struct are each stored in their own storage slot regardless of their size. This makes accessing or writing a particular struct field much more efficient.
+
+> **Note**
+> Struct variables are laid out in the same way today.
+
+For example, the storage variable `str` above is a string containing `99` characters and its size is `99` bytes which span 2 storage slots. It is also assigned index `2`. Therefore, `str` spans slots with keys `sha256("storage_2")` and `sha256("storage_2") + 1`.
+
+Variable `s` on the other hand, is a nested struct and is assigned index `3`. Therefore, each of its fields (and subfields) is assigned a separate key starting with `sha256("storage_3")`:
+
+- `s.a` is assigned `sha256("storage_3")`.
+- `s.b.c` is assigned `sha256("storage_3") + 1`.
+- `s.b.d` is assigned `sha256("storage_3") + 2`.
```

Comment:
OK, a couple of questions here.  Firstly, won't a 99 byte string require 4 32-byte slots?  99 / 32 = 3.09

I don't think the each-struct-field-goes-in-its-own-slot thing will fly.  If I have a struct with 4 ints in storage and I want to read and use them all, I am going to want them to be a single storage read and not to waste gas on 4 separate reads.  Isn't this why the Eth devs pack everything so tightly all the time?

If anything, putting each field into its own slot can only ever be _less_ efficent assuming that storage reads and writes are more expensive than a shift and a mask.  So I guess that is the question.

Reply:
And in the example at the top you show ` let s:Option<MyStruct> = s_handler.get();` which by the splitting logic would actually be two storage reads?

Equivalent to:
```rust
storage.s.x.get().and_then(|x| (x, storage.s.y.get())).and_then(|(x, y)| struct MyStruct { x, y })
```

Reply:
>  Firstly, won't a 99 byte string require 4 32-byte slots? 99 / 32 = 3.09

Lol yes of course

As for the struct splitting: 
1. This is how we do it today so this is not a regression. Not to say this justifies doing it that way.
2. One issue with packing is that if I want to update one internet field in a struct, I may need to read the whole slot where the field lives, update the 8 bytes where the integer is stored, and then store the whole thing back (read-modify-write). There is also the risk that some fields may now span 2 slots (e.g.: for (u64, b256), the b256 will spill over to the next slot, so updating it requires two read-modify-writes). In any case, the idea is that accessing or updating a single field is more efficient we the approach described here, but picking is more efficient when accessing or updating the whole struct.

Reply:
Ahh, right, read-modify-write is a thing.  We might need to offer both packed and unpacked options in the future, depending on what enough of the devs end up wanting.

So every leaf in a collection of nested structs gets its own slot?  And then they all get sequential keys based on declaration order?  If so we need to make sure that order is absolutely preserved thoughout the AST and beyond.

Reply:
On slot packing, the way solidity handles this is it starts at the first variable and if the current and following variables' sizes are small enough to fit into one slot, it will always be packed. Value types are never split between slots.

```solidity
contract A {
    // slot0[31]
    uint8 a;
    // slot0[32]
    uint8 b;
    // slot1[..]
    uint256 c;
    // slot2[12..]
    address d;
    // slot2[..12]
    uint96 e;
}
```

Since storage reads/writes are generally much more expensive than masking/shifting, in the future it would be good if packed was the default and padded was an option, possibly via a `#[padded]` attribute.

Reply:
Packing is now the default, at least for structs. This offers the most flexibility right now. If you want your variables to be packed, put them in a struct. Otherwise, leave them as their own storage variables.